ext/hash/hash.c

```diff
@@ -1170,6 +1362,145 @@ static zend_object *php_hashcontext_clone(zend_object *zobj) {
 }
 /* }}} */
 
+/* Serialization format: 5- or 6-element array
+   Index 0: hash algorithm (string)
+   Index 1: options (long, 0 or PHP_HASH_HMAC)
+   Index 2: hash-determined serialization of internal state (mixed, usually string)
+   Index 3: magic number defining layout of internal state (long)
+   Index 4: properties (array)
+   Index 5: key (string, only if PHP_HASH_HMAC)
+
+   HashContext serializations are not necessarily portable between architectures or
+   PHP versions. If the format of a serialized hash context changes, that should
+   be reflected in either a different value of `magic` or a different length of
+   the serialized context string. A particular hash algorithm can make its
+   HashContext serialization portable by parsing different representations in
+   its custom `hash_unserialize` method. */
+
+/* {{{ proto array HashContext::__serialize()
+   Serialize the object */
+PHP_METHOD(HashContext, __serialize)
+{
+	zval *object = ZEND_THIS;
+	php_hashcontext_object *hash = php_hashcontext_from_object(Z_OBJ_P(object));
+	zend_long magic = 0;
+	zval tmp;
+
+	if (zend_parse_parameters_none() == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	array_init(return_value);
+
+	if (!hash->ops->hash_serialize) {
+		goto serialize_failure;
+	}
+
+	ZVAL_STRING(&tmp, hash->ops->algo);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	ZVAL_LONG(&tmp, hash->options);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	if (hash->ops->hash_serialize(hash, &magic, &tmp) != SUCCESS) {
+		goto serialize_failure;
+	}
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	ZVAL_LONG(&tmp, magic);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	/* members */
+	ZVAL_ARR(&tmp, zend_std_get_properties(&hash->std));
+	Z_TRY_ADDREF(tmp);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+
+	if (hash->options & PHP_HASH_HMAC) {
+		ZVAL_STRINGL(&tmp, (const char *) hash->key, hash->ops->block_size);
+		zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &tmp);
+	}
+
+	return;
+
+serialize_failure:
+	zend_throw_exception_ex(spl_ce_UnexpectedValueException, 0, "HashContext for algorithm '%s' cannot be serialized", hash->ops->algo);
+	RETURN_THROWS();
+}
+/* }}} */
+
+/* {{{ proto void HashContext::__unserialize(array serialized)
+ * unserialize the object
+ */
+PHP_METHOD(HashContext, __unserialize)
+{
+	zval *object = ZEND_THIS;
+	php_hashcontext_object *hash = php_hashcontext_from_object(Z_OBJ_P(object));
+	HashTable *data;
+	zval *algo_zv, *magic_zv, *options_zv, *hash_zv, *members_zv, *key_zv;
+	zend_long magic, options;
+	const php_hash_ops *ops;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "h", &data) == FAILURE) {
```

Comment:
This is all good, Sara most probably wasn't aware that the zend_parse_parameters and the version with the throws suffix do the same thing in master (it wasn't the case for PHP 7.4)

Reply:
Off-topic, but do you know why this changed?

Reply:
> 
> 
> Off-topic, but do you know why this changed?

Pretty sure its due to: https://wiki.php.net/rfc/consistent_type_errors

Reply:
Yep, George is right :) Haha, @sgolemon is the only one who voted no for this RFC ^^